// music-metadata/lib/ogg/opus/OpusParser.js
'use strict';
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    }
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var Opus = require("./Opus");
var Token = require("token-types");
var VorbisParser_1 = require("../vorbis/VorbisParser");
/**
 * Opus parser
 * Internet Engineering Task Force (IETF) - RFC 6716
 * Used by OggParser
 */
var OpusParser = /** @class */ (function (_super) {
    __extends(OpusParser, _super);
    function OpusParser(metadata, options, tokenizer) {
        var _this = _super.call(this, metadata, options) || this;
        _this.tokenizer = tokenizer;
        _this.codecName = 'Opus';
        _this.lastPos = -1;
        return _this;
    }
    /**
     * Parse first Opus Ogg page
     * @param {IPageHeader} header
     * @param {Buffer} pageData
     */
    OpusParser.prototype.parseFirstPage = function (header, pageData) {
        // Parse Opus ID Header
        this.idHeader = new Opus.IdHeader(pageData.length).get(pageData, 0);
        if (this.idHeader.magicSignature !== "OpusHead")
            throw new Error("Illegal ogg/Opus magic-signature");
        this.metadata.setFormat('sampleRate', this.idHeader.inputSampleRate);
        this.metadata.setFormat('numberOfChannels', this.idHeader.channelCount);
    };
    OpusParser.prototype.parseFullPage = function (pageData) {
        var magicSignature = new Token.StringType(8, 'ascii').get(pageData, 0);
        switch (magicSignature) {
            case 'OpusTags':
                this.parseUserCommentList(pageData, 8);
                this.lastPos = this.tokenizer.position;
                break;
            default:
                break;
        }
    };
    OpusParser.prototype.calculateDuration = function (header) {
        if (this.metadata.format.sampleRate && header.absoluteGranulePosition >= 0) {
            // Calculate duration
            this.metadata.setFormat('numberOfSamples', header.absoluteGranulePosition - this.idHeader.preSkip);
            this.metadata.setFormat('duration', this.metadata.format.numberOfSamples / this.idHeader.inputSampleRate);
            if (this.lastPos !== -1 && this.tokenizer.fileSize && this.metadata.format.duration) {
                var dataSize = this.tokenizer.fileSize - this.lastPos;
                this.metadata.setFormat('bitrate', 8 * dataSize / this.metadata.format.duration);
            }
        }
    };
    return OpusParser;
}(VorbisParser_1.VorbisParser));
exports.OpusParser = OpusParser;
