// music-metadata/lib/common/GenericTagMapper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ID3v1Parser_1 = require("../id3v1/ID3v1Parser");
var CommonTagMapper = /** @class */ (function () {
    function CommonTagMapper(tagTypes, tagMap) {
        this.tagTypes = tagTypes;
        this.tagMap = tagMap;
    }
    CommonTagMapper.parseGenre = function (origVal) {
        // match everything inside parentheses
        var split = origVal.trim().split(/\((.*?)\)/g).filter(function (val) {
            return val !== '';
        });
        var array = [];
        for (var _i = 0, split_1 = split; _i < split_1.length; _i++) {
            var cur = split_1[_i];
            if (/^\d+$/.test(cur) && !isNaN(parseInt(cur, 10))) {
                cur = ID3v1Parser_1.Genres[cur];
            }
            array.push(cur);
        }
        return array
            .filter(function (val) {
            return val !== undefined;
        }).join('/');
    };
    CommonTagMapper.fixPictureMimeType = function (pictureType) {
        pictureType = pictureType.toLocaleLowerCase();
        switch (pictureType) {
            case 'image/jpg':
                return 'image/jpeg'; // ToDo: register warning
        }
        return pictureType;
    };
    CommonTagMapper.toIntOrNull = function (str) {
        var cleaned = parseInt(str, 10);
        return isNaN(cleaned) ? null : cleaned;
    };
    // TODO: a string of 1of1 would fail to be converted
    // converts 1/10 to no : 1, of : 10
    // or 1 to no : 1, of : 0
    CommonTagMapper.normalizeTrack = function (origVal) {
        var split = origVal.toString().split('/');
        return {
            no: parseInt(split[0], 10) || null,
            of: parseInt(split[1], 10) || null
        };
    };
    /**
     * Process and set common tags
     * @param comTags Target metadata to
     * write common tags to
     * @param tag     Native tag
     * @param value   Native tag value
     * @return common name
     */
    CommonTagMapper.prototype.mapGenericTag = function (tag) {
        tag = { id: tag.id, value: tag.value }; // clone object
        this.postMap(tag);
        // Convert native tag event to generic 'alias' tag
        var id = this.getCommonName(tag.id);
        return id ? { id: id, value: tag.value } : null;
    };
    /**
     * Convert native tag key to common tag key
     * @tag  Native header tag
     * @return common tag name (alias)
     */
    CommonTagMapper.prototype.getCommonName = function (tag) {
        return this.tagMap[tag];
    };
    /**
     * Handle post mapping exceptions / correction
     * @param {string} tag Tag e.g. {"©alb", "Buena Vista Social Club")
     */
    CommonTagMapper.prototype.postMap = function (tag) {
        return;
    };
    CommonTagMapper.maxRatingScore = 1;
    return CommonTagMapper;
}());
exports.CommonTagMapper = CommonTagMapper;
